Init script to extract survey responses

eljamm opened the pull request:
You can use [watchexec](https://search.nixos.org/packages?channel=unstable&show=watchexec&from=0&size=50&sort=relevance&type=packages&query=watchexec) to rapidly iterate on this as files change:

```sh
$ cd admin/scripts
$ watchexec 'python extract_survey_results.py survey_responses.json'
```

Closes: https://github.com/ngi-nix/summer-of-nix/issues/150

eljamm committed:
feat: init survey form models

eljamm committed:
feat: create alias mapping for questions

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,88 @@
+from pydantic import BaseModel, Field, field_validator, model_validator
+
+
+class Choice(str, Enum):
+    yes = "Yes"
+    not_planned = "Not planned"
+    not_yet = "Not yet"
+
+
+class Choice2(str, Enum):
+    yes = "Yes"
+    not_sure_yet = "Not sure yet"
+    no = "No"
```
Does it make sense to have these as an `Option[bool]`?

- yes -> True
- not_planned/not_sure -> None
- no/not_yet -> False

eljamm committed:
refactor: improve field naming

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,88 @@
+
+    name: str
+    author: Author
+    contributors: str
```
It would make more sense for this to be an `int`, but all alphanumeric characters are allowed in `q6`, so it's not a guarantee that it contains an integer.

fricklerhandwerk replied:
Make the datamodel what we need it to be, report parse errors when processing input

eljamm committed:
feat: map duration; dep update frequency

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,105 @@
+question_alias_mapping = {
+    "q1": "project_name",
+    "q2": "author_role",
+    "q3": "build_failure_duration",
```
There is no uniform formatting for this question. Some responses provide the number of hours per week directly (which we can validate and format as we want) while other go into more details.

Does it make sense to restrict the response field to numbers only?

fricklerhandwerk replied:
Let's keep it a string for now, this is a qualitative metric we need at the moment for prioritisation, but we likely won't have to work with it programmatically. Unless you think we do, then make it a number in a suitable unit, and we'll clean the data accordingly.

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    other = "Other"
+
+
+class UpdateFrequency(str, Enum):
+    cant_say = "Can't say"
+    more_once_per_day = "More than once per day"
+    more_once_per_week = "More than once per week"
+    more_once_per_month = "More than once per month"
+    more_once_per_year = "More than once per year"
+    less_once_per_year = "Less than once per year"
```
If we can fix a unit for this (say `X times per Y`), we could perhaps validate and convert it directly in the field.

fricklerhandwerk replied:
If one wants to be really rigorous, these answer options can only be expressed as a dependent type: `don't know | interval x, where x < Y | more than Y`. It can be done dynamically, but then you need to check that invariant at instance construction, which is probably more work than it's worth. Let's keep it an enum for now.

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,105 @@
+
+
+question_alias_mapping = {
+    "q1": "project_name",
```
I think there needs to be a distinction here about the project name. Right now, the question asks for the name as found on the [NLnet project list](https://nlnet.nl/project), but sometimes the name that the search bar returns does not correspond to the actual subgrant name. Rather, the subgrant name is found in the URL itself.

Perhaps the question needs to be reframed to make this distinction more clear, else we'll have a harder time mapping project names.

eljamm commented on admin/scripts/common/models/__init__.py:
```diff
@@ -0,0 +1,105 @@
+                    values[field_name] = values.pop(alias)
+            return values
+
+        @field_validator(question_alias_mapping["q24"], mode="before")
+        def map_yes_no_to_bool(cls, value: str) -> bool:
+            if value.lower() == "yes":
+                return True
+            return False
+
+        @field_validator(question_alias_mapping["q3"], mode="before")
+        def map_duration(cls, value: str) -> str:
+            if value.replace(".", "", 1).isdigit():
+                return f"{value}h per week"
+            return value
```
Does it make more sense the map values here (from the information source) or in the other data classes? For example, a `Project` might need the duration to be in a specific form, while `Notion` might require it to be in another. In that case, it would probably be better for field validation to happen separately.

eljamm committed:
docs: add note about the alias mapping

eljamm committed:
style: make enums in caps

eljamm committed:
refactor: rename ProjectRole to AuthorRole

It's the author's role within the project

eljamm committed:
feat: return invalid entries as errors

eljamm committed:
refactor: remove validator for duration

We probably don't need this for now

eljamm committed:
feat: add extensions field; validate multiple aliases

using the same validator and rename the alias map to be more concise.

eljamm committed:
refactor: put ci_cd in infra class

eljamm committed:
feat: init function to map project from response

eljamm committed:
feat: add q7, q8, q9

eljamm committed:
feat: init project from dict

eljamm committed:
refactor: devenv

eljamm committed:
feat: init nix ci_cd

eljamm committed:
feat: init nix fields

eljamm committed:
feat: add all remaining questions

eljamm committed:
refactor: rename mappings for better grouping

eljamm committed:
feat: remove alias map; group form fields

The alias map just adds redundancy and will be more difficult to
maintain. After further research, it should be possible to access the
field's alias and thus be able to know which question it came from (if
we ever need to).

eljamm committed:
feat: add contact info; available for pairing

eljamm committed:
feat: don't hardcode field names with a string

This wasn't a good idea to begin with, and it didn't really improve
anything. It just made it harder to spot when a field name changed.

eljamm committed:
feat: init project meta field

eljamm committed:
feaat: modify author

eljamm committed:
feat: longevity fields and release period

eljamm committed:
feat: don't use enums in project types

Not only are the field types already checked in the Form model, making
this redundant, but we also want to specify the types we want from the
project model, instead of being constrained to the form model.

eljamm committed:
fix: use Sequence for author role and channels

Apparently, a `list` does not handle converting from an enum to a
string, but a `Sequence` does, since it's a more general collection of
data.

eljamm committed:
feat: author responses; dependency and framework

eljamm committed:
feat: add artefacts and documentation
